OrbitLinuxTracing/UprobesReturnAddressManagerTest.cpp

```diff
@@ -0,0 +1,393 @@
+            test_handler.GetExpectedStack());
+
+  // D returns.
+  test_handler.OnNonUretprobesReturn();
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // C returns and hits a uretprobe.
+  test_handler.OnUretprobesReturn(&return_address_manager);
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // B returns and hits a uretprobe.
+  test_handler.OnUretprobesReturn(&return_address_manager);
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // E is called and hits a uprobe.
+  test_handler.OnUprobesCall(&return_address_manager);
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // E returns and hits a uretprobe.
+  test_handler.OnUretprobesReturn(&return_address_manager);
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // A returns.
+  test_handler.OnNonUretprobesReturn();
+}
+
+TEST(UprobesReturnAddressManager, TailCallOptimization) {
+  UprobesReturnAddressManager return_address_manager;
+  TestHandler test_handler{42};
+
+  // A is called.
+  test_handler.OnNonUprobesCall();
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // B is called and hits a uprobe.
+  test_handler.OnUprobesCall(&return_address_manager);
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // C is called with tail-call optimization and hits a uprobe.
+  test_handler.OnUprobesOptimizedTailCall(&return_address_manager);
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // D is called.
+  test_handler.OnNonUprobesCall();
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // D returns.
+  test_handler.OnNonUretprobesReturn();
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // C returns and hits a uretprobe.
+  test_handler.OnUretprobesReturn(&return_address_manager);
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // B is not on the stack anymore because it had ended with a tail-call, but
+  // its uretprobe is still hit.
+  test_handler.OnUretprobesAfterTailCallReturn(&return_address_manager);
+
+  EXPECT_EQ(test_handler.PatchStackOnSample(&return_address_manager),
+            test_handler.GetExpectedStack());
+
+  // A returns.
+  test_handler.OnNonUretprobesReturn();
+}
+
+}  // namespace LinuxTracing
```

Comment:
Just a suggestion for the future commits. It would be easier to read if the test and Implementation of ReturnAddressManager was in a separate commit and switch from old way of fixing call-stacks to new one in follow up commit.

Reply:
Good point, I'll keep that in mind.